voicewatch/voicewatch.py

```diff
@@ -0,0 +1,98 @@
+from datetime import datetime, timezone, timedelta
+from typing import Optional
+
+import discord
+from redbot.core import Config, checks
+from redbot.core.bot import Red
+from redbot.core import commands
+
+
+class VoiceWatchCog(commands.Cog):
+    """VoiceWatch Cog"""
+
+    def __init__(self, bot: Red):
+        self.bot = bot
+        self.config = Config.get_conf(self, identifier=128986274420752384003)
+        self.alertCache = {}
+        default_guild_config = {
+            "logchannel": "",  # Channel to send alerts to
+            "min_joined_hours": 8760  # Default to one year so the mods set it up :)
+        }
+
+        self.config.register_guild(**default_guild_config)
+
+    @checks.admin()
+    @commands.group("voicewatch", aliases=["vw"], pass_context=True)
+    async def _voicewatch(self, ctx: commands.Context):
+        pass
+
+    @_voicewatch.command(name="time")
+    async def _time(self, ctx: commands.Context, hours: str):
+        """Set/update the minimum hours users must be in the server before without triggering an alert."""
+        try:
+            hrs = int(hours)
+            await self.config.guild(ctx.guild).min_joined_hours.set(hrs)
+            await ctx.send("✅ Time requirement successfully updated!")
+        except ValueError:
+            await ctx.send("Error: Non-number hour argument supplied.")
+
+    @_voicewatch.command(name="logchannel")
+    async def _logchannel(self, ctx: commands.Context, channel: Optional[discord.TextChannel]):
+        """Set/update the channel to send voice activity alerts to."""
+
+        chanId = ctx.channel.id
+        if channel:
+            chanId = channel.id
+        await self.config.guild(ctx.guild).logchannel.set(chanId)
+        await ctx.send("✅ Alert channel successfully updated!")
+
+    @commands.Cog.listener()
+    async def on_voice_state_update(self, member: discord.Member,
+                                     before: discord.VoiceState, after: discord.VoiceState):
+        if member.id in self.alertCache and self.alertCache[member.id]:  # Ignore if we've already alerted on this user. Can probably find a better way.
+            # Todo: Make guild-aware
+            return
+        else:
+            self.alertCache[member.id] = False
+
+        if after is None or after.channel is None:  # Check if we're missing the after data or associated channel.
```

Comment:
> `if after is None`

Is this a possible case? The docs don't indicate that it is, and as it's typed as `VoiceState` and not `Optional[VoiceState]` I wouldn't expect `after` to ever be `None`.

Reply:
I don't remember details but I know I've had issues with this in the past (may have been with a different framework) so its just a harmless safety check at this point. It can be removed if you think its necessary, but I'd prefer to leave it otherwise.